api/src/main/java/org/apache/apex/api/Launcher.java

```diff
@@ -0,0 +1,172 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.api;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Iterator;
+import java.util.ServiceLoader;
+
+import org.apache.hadoop.conf.Configuration;
+
+import com.google.common.base.Throwables;
+
+import com.datatorrent.api.Attribute;
+import com.datatorrent.api.StreamingApplication;
+
+/**
+ *
+ */
+public abstract class Launcher<H extends Launcher.AppHandle>
+{
+
+  public static final String NEW_INSTANCE_METHOD = "newInstance";
+
+  public static class LaunchMode<L extends Launcher>
+  {
+    public static final LaunchMode<EmbeddedAppLauncher> EMBEDDED = new LaunchMode<>(EmbeddedAppLauncher.class);
+    public static final LaunchMode<YarnAppLauncher> YARN = new LaunchMode<>(YarnAppLauncher.class);
+
+    Class<L> clazz;
+
+    public LaunchMode(Class<L> clazz)
+    {
+      this.clazz = clazz;
+    }
+  }
+
+  public enum ShutdownMode
+  {
+    ORDERLY, KILL
```

Comment:
I think we should use AWAIT_TERMINATION and KILL/IMMEDIATE or something similar that is already established.